core/src/main/java/org/jdbi/v3/core/mapper/ConstructorMapper.java

```diff
@@ -69,32 +69,36 @@ public T map(ResultSet rs, StatementContext ctx)
     }
 
     private ObjectCreator<T> createCreator(List<String> columnNames, ResultSetMetaData metadata, StatementContext ctx) {
-        final int length = columnNames.size();
+        final int length = constructor.getParameterCount();
 
-        if (length != constructor.getParameterCount()) {
-            throw new IllegalStateException(length + " columns in result set, but constructor takes " + constructor.getParameterCount());
+        if (length > columnNames.size()) {
+            throw new IllegalStateException(columnNames.size() +
+                    " columns in result set, but constructor takes " +
+                    constructor.getParameterCount());
         }
 
         final int[] columnMap = new int[length];
         final ColumnMapper<?>[] mappers = new ColumnMapper<?>[length];
 
         for (int i = 0; i < length; i++) {
             final Type type = constructor.getGenericParameterTypes()[i];
-            final int paramIndex = parameterIndexForColumn(columnNames.get(i));
+            final String paramName = paramName(constructor.getParameters()[i]);
+            final int columnIndex = columnIndexForParameter(columnNames, paramName);
 
-            if (mappers[paramIndex] != null) {
-                throw new IllegalArgumentException("Column named " + columnNames.get(i) + " maps to multiple parameters");
+            if (mappers[i] != null) {
+                throw new IllegalArgumentException("Parameter named " +
+                        paramName + " maps to multiple columns");
```

Comment:
Since we're now iterating through the constructor parameters looking for matching columns (instead of the other way around), I believe this if-block is moot now.